compiler/damlc/daml-visual/src/DA/Daml/Visual.hs

```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
 addCreateChoice TemplateChoices {..} lookupData = nodeIdForChoice lookupData tplNameCreateChoice
     where tplNameCreateChoice = LF.ChoiceName $ T.pack $ DAP.renderPretty (headNote "addCreateChoice" (LF.unTypeConName (LF.tplTypeCon template))) ++ "_Create"
 
-constructSubgraphsWithLables :: Map.Map LF.ChoiceName ChoiceDetails -> TemplateChoices -> SubGraph
-constructSubgraphsWithLables lookupData tpla@TemplateChoices {..} = SubGraph nodesWithCreate template
+-- This is copied from PrettyScenarios but depending on SS for visual seems odd
+labledField :: T.Text -> T.Text -> T.Text
+labledField fname "" = fname
+labledField fname label = fname <> "." <> label
+
+typeConFieldsNames :: LF.World -> (LF.FieldName, LF.Type) -> [T.Text]
+typeConFieldsNames world (LF.FieldName fName, LF.TConApp tcn _) = map (labledField fName) (typeConFields tcn world)
+typeConFieldsNames _ (LF.FieldName fName, _) = [fName]
+
+typeConFields :: LF.Qualified LF.TypeConName -> LF.World -> [T.Text]
+typeConFields qName world = case LF.lookupDataType qName world of
+  Right dataType -> case LF.dataCons dataType of
+    LF.DataRecord re -> concatMap (typeConFieldsNames world) re
+    LF.DataVariant _ -> [""]
+    LF.DataEnum _ -> [""]
+  Left _ -> error "malformed template constructor"
+
+-- Other way is to go from Expr to Qalified a and then use that qualified
+tplQulified :: LF.World -> TemplateChoices -> [T.Text]
```

Comment:
This function has a very confusing name. It sounds like it somehow returns the qualified name of a template but that’s not what it does. It returns a list of field names. It might be easier to understand this code if you work with `LF.FieldName` everywhere and only convert to `Text` at the very end.